File: Chapter1/multiprocessing/multiprocess.jl
#! /usr/bin/env julia

using Distributed

# Add processes (if not started with `julia -p n`)
addprocs(10) 
# addprocs([("localhost", 2), ("astro-node10", 2)], topology=:master_worker)

# Define a function to square a number
@everywhere function square(number)
    return number * number
end

# Example function to precompile 'square'
@everywhere begin
    square(1)
end

# Define a range of numbers
numbers = 0:1_000 # Define the range of numbers

# Define a function to calculate squares
function calculate_squares(numbers)
    return pmap(square, numbers)
end

# Timing the execution
@time squared_numbers = calculate_squares(numbers)

println(squared_numbers[1:5],"...",squared_numbers[end-4:end])

File: Chapter1/multiprocessing/multiprocess.py
#! /usr/bin/env python

# Calculate squares in parallel

# Import packages
from multiprocessing import Pool

# Square number
def square(number):
    return number * number

# Calculate squares of numbers in parallel
def calculate_squares(numbers):

    # Create a pool of workers
    with Pool(processes=10) as pool:
        # Map the 'square' function to the list of numbers
        chunksize = 10 # The number of elements per chunk, very important!
        results = pool.map_async(square, numbers, chunksize)
        pool.close()
        pool.join()
    return results

# Main function
if __name__ == "__main__":

    # Define a list of numbers
    numbers = range(1_000+1) 
    squared_numbers = calculate_squares(numbers)
    print(squared_numbers[:5],'...',squared_numbers[-5:])